### PR TITLE
Reset the YOLOE predictor after the vocabulary delegate

### PR DESCRIPTION
**Pull request by raimbekovm**
## summary

yoloe.set_vocab and get_vocab cleared the cached predictor before delegating, so every assert inside the delegate raised with model.predictor already dropped, and the next predict silently rebuilt the default predictor from overrides instead of the custom predictor class the caller had passed once and was reusing. the reset now runs after the delegate returns, so a rejected call leaves the model exactly as it found it.

## measured

on yoloe-26n-seg.pt with a predictor warmed by one predict call, both rejected routes previously left predictor before=true after=false: get_vocab on an already fused head, and set_vocab on a model put back in train mode. after the change both keep the same predictor object. a successful get_vocab and set_vocab still clear it and predict reports the new class names, byte identical to the base run.

the automated review asks whether preserving the predictor after a rejected set_vocab can leave a train mode model behind. measured as a 2x2 over trigger and revision: predict, then model.model.train(), then predict already fails with keyerror 0 in segment postprocess on current main with no set_vocab in the picture, so that hazard is pre existing and belongs to the predictor, not here. what this change removes is the accident that hid it on one route. filed separately with the table and a fix at the owner.

stacked on #25744, which is the other open change to this file.

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary

YOLOE now clears its cached predictor only after `set_vocab` or `get_vocab` succeeds, preserving the existing predictor when either delegated operation is rejected.

### 📊 Key Changes

- Moved predictor reset in `set_vocab` to after `self.model.set_vocab(...)` returns.
- Updated `get_vocab` to store the delegated result, clear the predictor after success, and then return the result.
- Rejected vocabulary operations no longer discard the current predictor before raising an assertion.
- Successful vocabulary operations continue to invalidate the cached predictor.

### 🎯 Purpose & Impact

- A failed `set_vocab` or `get_vocab` leaves the model and its existing predictor unchanged, including custom predictor instances.
- Successful vocabulary updates still cause the next prediction to rebuild the predictor and use the updated class names.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @raimbekovm, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This automated message confirms your contribution was received, and an Ultralytics engineer will assist with the review. To ensure a seamless integration of your work, please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Reviewed the YOLOE vocabulary reset ordering and its predictor lifecycle. The post-delegate reset fixes the reported loss of custom predictors on validation failures, but the rejected `set_vocab()` training-mode path can leave a reused predictor wrapping a training-mode model, causing subsequent prediction failures or incorrect outputs.

<details><summary>💬 Posted 1 inline comment</summary>

- 💡 **MEDIUM** `ultralytics/models/yolo/model.py:324` This preserves a cached predictor even when `set_vocab()` is rejected because the underlying model is in training mode. If the caller then invokes `predict()` without first restoring eval mode, `Model.predict()` reuses the existing predictor and skips `setup_model()`, which is the only path here that calls `eval()`. The predictor wraps the same raw module, so YOLOE can run its training-mode forward and return training-form outputs instead of inference predictions. Ensure a reused predictor re-e…
</details>

File: ultralytics/models/yolo/model.py [hunk 1]
```diff
@@ -320,14 +320,15 @@ def set_vocab(self, vocab: torch.nn.ModuleList, names: list[str]) -> None:
         """
         assert isinstance(self.model, YOLOEModel)
         names = check_class_names(names)
-        self.predictor = None  # the delegate destructively re-parameterizes the head
         self.model.set_vocab(vocab, names=names)
+        self.predictor = None  # after the delegate, so a rejected call leaves the predictor alone
```
Comment (UltralyticsAssistant) [inline ultralytics/models/yolo/model.py:324]: 💡 **MEDIUM**: This preserves a cached predictor even when `set_vocab()` is rejected because the underlying model is in training mode. If the caller then invokes `predict()` without first restoring eval mode, `Model.predict()` reuses the existing predictor and skips `setup_model()`, which is the only path here that calls `eval()`. The predictor wraps the same raw module, so YOLOE can run its training-mode forward and return training-form outputs instead of inference predictions. Ensure a reused predictor re-enters eval mode before inference, or otherwise handle this rejected-call state without leaving a train-mode model behind.

**Comment by glenn-jocher** [comment]
Closing because the proposed cache-preservation change creates a real inference regression. I reproduced the reported rejected-call state with a live YOLOE model and an existing predictor: on main the predictor is cleared, the next predict rebuilds it, calls eval(), and succeeds; on this head the cached predictor survives while the model remains in train mode, and the next predict fails in BatchNorm. Clearing a derived predictor on a rejected destructive delegate is safe cache invalidation; preserving it is not. The unresolved review thread correctly identifies this behavior.
